### ISSUE TITLE
Question on syslog over tls handshake failure scenario.

### ISSUE THREAD

**Issue by Anusha-Pai95** [rsyslog/rsyslog#4503]
on a syslog over tls client, which api can help us to get the tls version enabled in server . In order to detect and capture a failure handshake scenario in rsyslog tls client when version mismatched between client and server. For example syslog tls client running on tls1.2 and openssl server supporting tls1 which results in version mismatch between syslog tls client and server. Inorder to capture this scenaio do we have any predetermined api that helps us to get the version enabled on server from the client.

**Comment by Anusha-Pai95** [comment]
wanted to capture the this error condition by inserting a log in the client during this handshake failure case.

**Comment by Anusha-Pai95** [comment]
there is a function SSL_get_version which return the protocol version and it requires ssl object. is there any way to get the server ssl object while running in the client. So server's ssl object can be passed to the above mentioned function to get the server openssl protocol version and it if is unsupported in the client then adding a log to capture this failure condition could actually solve this problem.

**Comment by rgerhards** [comment]
I do not understand your intent. Do you plan to make a modification to rsyslog?

**Comment by davidelang** [comment]
I think his intent is to be able to have a log that tells him why the handshake 
failed

However, in the case where the encryption settings don't match up, I suspect 
that you don't get the details of what settings would work, just tht nothing 
matched the config. Only if it negotiates successfully will it give details on 
what encryption was used.

David Lang

**Comment by Anusha-Pai95** [comment]
intent was to get a log indicating the reason for syslog over tls handshake failure. In case of version incompatibility between the client and server, how can this case be detected from syslog tls client.

**Comment by davidelang** [comment]
it really can't. all the client knows is that the server did not accept the 
connection, it doesn't know why the server did it.

you basically need to have debug level logging to see exactly what's being 
passed over the wire in each direction to begin to figure out the why (or try to 
connect with more permissive settings and see what the resulting connection is 
if it's accepted)

**Comment by rgerhards** [comment]
Apologies for the very late reply. This may no longer be relevant for the original reporter, but I am adding the answer for people who find this issue later via search. We do not lightly close old issues, and we are working through the backlog rather than closing issues just because of age.

The short answer is that a TLS client generally cannot ask the server for a complete list of acceptable protocol versions after a failed handshake. If the protocol/cipher settings do not overlap, the handshake fails before there is a successfully negotiated TLS session. Functions such as `SSL_get_version()` can report the negotiated protocol version only after negotiation succeeds; they cannot reveal the server's full policy from a failed client-side handshake.

For troubleshooting version mismatch cases, the practical approaches are:

- inspect rsyslog's own TLS error message and the peer/server logs
- test with `openssl s_client` using explicit `-tls1_2`, `-tls1_3`, cipher, and CA/cert options
- temporarily test with more permissive settings to see what version/cipher gets negotiated, then tighten policy again
- use packet capture/debug logging when the failure reason must be proven

So this is not a missing rsyslog API that can reliably discover the server's TLS versions from the client after failure; it is a limitation of how TLS negotiation works. I am closing this as answered. If a current rsyslog log message is misleading for a specific protocol-version mismatch, please open a new issue with both endpoints' TLS policy and the exact error output.
